File: src/main/java/dev/macrohq/swiftslayer/pathfinding/PathExecutor.java
package dev.macrohq.swiftslayer.pathfinding;

import cc.polyfrost.oneconfig.config.annotations.KeyBind;
import dev.macrohq.swiftslayer.util.*;
import net.minecraft.client.settings.KeyBinding;
import net.minecraft.util.*;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.util.ArrayList;
import java.util.List;

public class PathExecutor {
    private List<BlockPos> path;
    public boolean running = false;
    private BlockPos current = null;
    public float directionYaw = 0f;

    public void executePath(List<BlockPos> inputPath) {
        if (running) return;
        path = new ArrayList<>(inputPath);
        current = path.get(0);
        running = true;
        directionYaw = Ref.player().rotationYaw;
    }

    @SubscribeEvent
    public void onTick(TickEvent.ClientTickEvent event) {
        if(Ref.world()==null || Ref.player()==null) return; // :angiest:
        if (!running) return;

        if(path.contains(PlayerUtil.getStandingPosition())){
            if(path.indexOf(PlayerUtil.getStandingPosition()) == path.size()-1){
                running = false;
                KeyBindUtil.setPressed(Ref.gameSettings().keyBindSprint, false);
                KeyBindUtil.setPressed(Ref.gameSettings().keyBindForward, false);
                KeyBindUtil.setPressed(Ref.gameSettings().keyBindJump, false);
                return;
            }
            current = path.get(path.indexOf(PlayerUtil.getStandingPosition())+1);
        }
        movePlayer(current);
    }

    public void disable(){
        running = false;
    }

    private void movePlayer(BlockPos current) {
        RotationUtil.Rotation rotation = RotationUtil.getAngles(new Vec3(current.getX() + 0.5, current.getY() + 2.0, current.getZ() + 0.5));
        RotationUtil.ease(rotation, 500);
        directionYaw = rotation.getYaw();
        KeyBindUtil.setPressed(Ref.gameSettings().keyBindSprint, true);
        KeyBindUtil.setPressed(Ref.gameSettings().keyBindForward, true);
        KeyBindUtil.setPressed(Ref.gameSettings().keyBindJump, current.getY() > Ref.player().posY - 1);
    }

    @SubscribeEvent
    public void onRenderWorldLast(RenderWorldLastEvent event) {
        RotationUtil.onRenderWorldLast();
    }
}

File: src/main/kotlin/dev/macrohq/swiftslayer/pathfinding/PathExecutor.kt
package dev.macrohq.swiftslayer.pathfinding

import dev.macrohq.swiftslayer.feature.helper.Angle
import dev.macrohq.swiftslayer.util.*
import dev.macrohq.swiftslayer.util.rotation.RotationManager
import me.kbrewster.eventbus.Subscribe
import net.minecraft.entity.EntityLiving
import net.minecraft.util.BlockPos
import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent
import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.sqrt

class PathExecutor {
  var enabled = false
    private set
  var directionYaw = 0f
    private set
  private var path = listOf<BlockPos>()
  private var next: BlockPos? = null
  private var pathFailCounter = 0
  private var aotving = false
  private var goal: BlockPos? = null
  private var rotate: Boolean = true

  @Subscribe
  fun onTick(event: ClientTickEvent) {
    if (!enabled) return
    if(RotationManager.getInstance().currentThread.isAlive) {
      gameSettings.keyBindSprint.setPressed(false)
      gameSettings.keyBindForward.setPressed(false)
      gameSettings.keyBindJump.setPressed(false)
      return
    }




    if (!isOnPath()) {
      if (pathFailCounter >= 100) {
        Logger.error("Path execution failed, retrying.")
        disable()
        PathingUtil.goto(path[path.size - 1])
        return
      }
      if (player.onGround) pathFailCounter++
    }
    pathFailCounter = 0

    if (isOnPath() && hasReachedEnd()) {
      disable()
      return
    }

    if (isOnPath()) {
      next = path[path.indexOf(getStandingOn()!!) + 1]
      if(rotate && !shouldJump()) {
        if (goal != null) {
        RotationManager.getInstance().rotateTo(goal!!.toVec3Top(), 2f)
        } else {
          RotationManager.getInstance().rotateTo(next!!.toVec3Top(), 2f)
        }
      }

      RenderUtil.markers.clear()
      RenderUtil.markers.add(next!!)
    }

    if (canAOTV()) {
      if (InventoryUtil.holdItem("Aspect of the Void")) {
        KeyBindUtil.rightClick()
        aotving = true
        return
      }
      Logger.error("No AOTV in hotbar.")
    }

    if (aotving && !hasAOTVed()) {
      Logger.log("Waiting for AOTV to arrive.")
      return
    } else if (aotving && hasAOTVed()) {
      Logger.log("AOTV Finished.")
      aotving = false
    }

   // val rotation = RotationUtil.Rotation(AngleUtil.getAngles(next!!.toVec3Top()).yaw, 20f)
    val rotation = Angle(AngleUtil.getAngle(next!!.toVec3Top()).yaw, 20f)
    directionYaw = rotation.yaw
   gameSettings.keyBindSprint.setPressed(AngleUtil.yawTo360(player.rotationYaw) in AngleUtil.yawTo360(player.rotationYaw) - 45..AngleUtil.yawTo360(player.rotationYaw) + 45)
    gameSettings.keyBindForward.setPressed(true)
    gameSettings.keyBindJump.setPressed(shouldJump())

  }

  fun enable(pathIn: List<BlockPos>, target: EntityLiving? = null, rotate: Boolean = true) {
    if (pathIn.isEmpty()) return
    disable()
    path = pathIn
    next = path[0]
    directionYaw = player.rotationYaw
    pathFailCounter = 0
    aotving = false
    enabled = true
    if(target != null) {
      goal = target.position
    }
    this.rotate = rotate

  }

  fun disable() {
    enabled = false
    gameSettings.keyBindSprint.setPressed(false)
    gameSettings.keyBindForward.setPressed(false)
    gameSettings.keyBindJump.setPressed(false)
    goal = null
    rotate = true
  }

  private fun canAOTV(): Boolean {
    val d = sqrt(player.getDistanceSqToCenter(next)) > 9 && !aotving && path.indexOf(next) != path.size - 1
    val yp = AngleUtil.getAngles(next!!.up().up())
    val yawDiff = abs(AngleUtil.yawTo360(player.rotationYaw) - AngleUtil.yawTo360(yp.yaw))
    val pitchDiff = abs(mc.thePlayer.rotationPitch - yp.pitch)
    return d && yawDiff < 5 && pitchDiff < 2 && false
  }

  private fun shouldJump() = player.onGround && (next!!.y + 0.5 - player.posY) >= 0.5 && (sqrt(
    (player.posX - next!!.x).pow(
      2.0
    ) + (player.posZ - next!!.z).pow(2.0)
  ) < 2)

  private fun isOnPath() = getStandingOn() != null
  private fun getStandingOn() = path.find { it.x == player.getStandingOnCeil().x && it.z == player.getStandingOnCeil().z }
  private fun hasReachedEnd() = player.getStandingOnCeil().x == path[path.size - 1].x && player.getStandingOnCeil().z == path[path.size - 1].z
  private fun hasAOTVed() = sqrt(player.lastTickPositionCeil().distanceSq(player.getStandingOnCeil())) > 4

}